### ISSUE TITLE
OVN performance needs to be optimized

### ISSUE THREAD

**Issue by zhanrox2** [ovn-org/ovn#50]
Test Environment

- OVN Version：20.03

- OVN_NB：cluster mode with 3 nodes

- OVN_SB：cluster mode with 3nodes

- OVN-Northd：active and standby mode with 3 nodes

- OVN-Controller：33 nodes

- OS：CentOS Linux release 8.0.1905 (Core)

- Kernel：4.18.0-80.el8.x86_64

**1. About ovn-northd process**
In order to find the changed data, each loop in the main function will load the full amount of data from ovnnb_db and ovnsb_db. When the amount of data is not large, such as created 2000 switches or 2000 ports, the CPU corresponding to the ovn-northd process is 100% busy, regardless of whether there are data changes or new operations. At this time, the CPUs of ovnnb_db and ovnsb_db have been kept at about 50%.
This mechanism of full load and full check has seriously affected the performance of ovn. It is recommended to change the mechanism of detecting incremental change data to reduce the amount of calculation and improve performance.

**2. About ovn-nbctl client tools**
Any command executed via ovn-nbctl will load the full amount of data from ovnnb_db to the client, regardless of whether the command requires these data. When the amount of data is large, this command will take a long time to execute.
When the number of ports under the switch reaches 9197, any addition command, such as ls-add, will report "database connection failed (Broken pipe)" error。

I want to know whether our OVN technical roadmap will make optimization in the above aspects。

**Comment by hzhou8** [comment]
Thanks for raising the problems.

For 1), there is a project converting ovn-northd to incremental processing using differential datalog (DDlog). https://github.com/ovn-org/ovn/tree/ddlog-dev-v2

For 2), there is already daemon mode supported by ovn-nbctl. See "man ovn-nbctl".

There has been many OVN performance and scale related discussions and improvements. Here are some examples:

https://www.slideshare.net/hanzhou1978/large-scale-overlay-networks-with-ovn-problems-and-solutions

https://www.openvswitch.org/support/ovscon2019/day2/1351-OVN%20Performance.pdf

**Comment by billowjen** [comment]
1), Is datalog a Long term stable version?  Will it be merged into the master branch?
2), daemon mode cannot resolve problem. zhanrox2's questions, test based on  daemon mode
3), We found that using jsonapi to operate ovsdb was very fast，so i think optimiz nbctl is more valuable

**Comment by zhanrox2** [comment]
@hzhou8  Thank you for your reply !
1), Is datalog a Long term stable version? Will it be merged into the master branch? But i not found which lines to incremental processing, I continue to study in depth.
2), daemon mode cannot resolve problem. my test just based on the daemon mode
3), We found that using jsonapi to operate ovsdb was very fast，so i think optimiz nbctl is more valuable

**Comment by rnurgaliyev** [comment]
Hello, I would like to tackle this one more time :)

As far as I can tell after reading mailing list, work on differential datalog got stuck after some point. If I understood it right, DDlog is preferred as a general approach to avoid full recomputation in ovn-northd and ovn-controller, so amazing work on incremental processing engine, done by @hzhou8, was never merged into master. We are now going to use OVN in scale, and would really love to avoid known issues. So, here are couple questions:

1. What can be done to unblock DDlog development? Any help or resources needed?

2. Is there a diff or a fork for incremental processing in ovn-controller for latest stable branch (20.09)? I was trying to find something in Mr. @hzhou8 repos but got lost there. Links to branches that are mentioned in slides are for way outdated code.

**Comment by blp** [comment]
I'm actively working toward getting the ddlog implementation merged. We have just tentatively agreed to freeze new features in the C version until the ddlog version is in. A few initial (documentation) patches from the ddlog implementation are already merged.

You can follow my patches for ddlog in my "reviews" repo (https://github.com/blp/ovs-reviews/branches). I'm using the ddlog# branches, where # increments as I revise things. And of course I'm posting series of patches as the previous series are reviewed and merged.

**Comment by rnurgaliyev** [comment]
@blp thank you very much for the reply! We will try our best to test your ddlog implementation and provide some feedback that could benefit development.

**Comment by blp** [comment]
Much appreciated!

**Comment by hzhou8** [comment]
@rnurgaliyev sorry that I didn't see your question until now. For your question 2), maybe you already got the answer but just for the record: ovn-controller incremental processing has been merged upstream back in the mid of 2019, so in the branch 20.09 you mentioned (and even before that) it is already there. The slides you were seeing was made just before the community decided to merge it upstream. So, sorry for the confusion. And for DDlog, so far there is still no concrete plan yet to use DDlog for ovn-controller.

**Comment by a821485769-coder** [comment]
@hzhou8 Is the latest version of northd still processing the data of the logical router in full?
